### PR TITLE
Additions to allocations enforcement

### PR DESCRIPTION
**Pull request by steve-gregory**
## Description
- [x] Simplify the logic for calculating remaining time/over allocation, storing all logic in AllocationSource
- [x] Test and verify that allocations are enforced on the backend and new instances cannot be launched, when the user has exhausted their allocation.
- [x] Change 'related_name' for user, Include a note about 'related_name' for allocation_source

### CODE REVIEW

File: core/models/allocation_source.py [hunk 1]
```diff
@@ -21,20 +21,42 @@ def for_user(cls, user):
...
+        time_shared_allocations = getattr(settings, 'SPECIAL_ALLOCATION_SOURCES', {})
+        if user and self.name in time_shared_allocations.keys():
+            try:
+                compute_allowed = time_shared_allocations[self.name]['compute_allowed']
+            except:
+                raise Exception(
+                    "The structure of settings.SPECIAL_ALLOCATION_SOURCES "
+                    "has changed! Verify your settings are correct and/or "
+                    "change the lines of code above.")
+            last_snapshot = self.user_allocation_snapshots.filter(user__username='test-sgregory').order_by('-updated').last()
```
Comment (julianpistorius) [inline core/models/allocation_source.py]: From our chat, there should only be one user_allocation_snapshot per user and allocation source because of this line in the `UserAllocationSnapshot` class:

```python
unique_together = ('user','allocation_source')
```

Check my logic, but I think this might do it:

```python
self.user_allocation_snapshots.get(user__username='test-sgregory')
```

Comment (julianpistorius) [reply core/models/allocation_source.py]: Also you'll probably want to use `user.username` instead of the hard-coded `test-sgregory` username here. ;)

File: api/v2/serializers/details/instance.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def get_size(self, obj):
         return serializer.data
 
     def get_image(self, obj):
-        if obj.source.is_volume():
+        if not obj.source.is_machine():
```
Comment (julianpistorius) [inline api/v2/serializers/details/instance.py]: Does the commit message mean that the full fix is coming in another PR? Should this be a separate PR?

File: core/models/allocation_source.py [hunk 3]
```diff
@@ -21,20 +21,42 @@ def for_user(cls, user):
     def get_instance_ids(self):
         return self.instanceallocationsourcesnapshot_set.all().values_list('instance__provider_alias', flat=True)
 
-    def check_over_allocation(self):
+    def is_over_allocation(self, user=None):
```
Comment (julianpistorius) [inline core/models/allocation_source.py:24]: Nice.

**Review by julianpistorius — CHANGES_REQUESTED** [review]
Remove hard-coded username, and see note about maybe splitting off ATMO-1880 mitigation into separate PR. Otherwise 👍

File: api/v2/serializers/details/instance.py (hunk 2 above)
Comment (steve-gregory) [reply api/v2/serializers/details/instance.py]: Yep, added the code to `yampy-yellowlegs` instead

**Review by julianpistorius — APPROVED** [review]
👍
